[Step3] 자동차 경주

y2gcoder opened the pull request:
안녕하십니까!

주말에도 제 코드를 봐주셔서 감사합니다. 
요구사항만 보고 가볍게 생각했는데 볼수록 제가 너무 어렵게 생각하는 것 같아 바로 리뷰 올리겠습니다! 
다시 한 번 감사합니다!

양영근 올림

y2gcoder committed:
feature(step3-racingcar): Add Car

자동차 경주 게임 중 자동차 클래스 구현.
랜덤한 숫자를 받아 조건에 따라 움직이는 move()에 대한 테스트 코드 작성.(Kotest로 작성)

y2gcoder committed:
test(step3-racingcar): Modify test's text: 자동차 -> Car

y2gcoder committed:
test(step3-racingcar): Create constant DEFAULT_POSITON in Car

매직넘버를 피하기 위해 Car 내부에 기본 위치에 대한 상수값을 지정.

y2gcoder committed:
test(step3-racingcar): Modify Implementation of Car's move()

if 가 너무 많아서 찾아보니 코틀린의 when() 이 있어 적용.

y2gcoder committed:
test(step3-racingcar): Add learning test about List Creation with element initialization

코틀린의 리스트 요소 초기화 방법에 대한 학습 테스트 작성

y2gcoder committed:
feature(step3-racingcar): Add RacingGame and RacingGameTest

레이싱 게임의 비즈니스 로직을 담고 있는 레이싱 게임 클래스 생성.
레이싱 게임의 결과를 List<RoundResult>의 형태로 반환하도록 함.

y2gcoder committed:
feature(step3-racingcar): Add main function, InputView and RacingGameRaceResultView

자동차 대수와 라운드 수(시도할 횟수)를 받기 위한 InputView 추가
레이스 결과를 발표해줄 RacingGameResultView 추가
끝으로 실행할 메인 function 파일도 추가

y2gcoder committed:
docs(step3-racingcar): Check step3 requirement

3단계 요구사항 구현 체크

ddaaac commented on src/main/kotlin/week1/step3/Car.kt:
```diff
@@ -0,0 +1,17 @@
+package week1.step3
+
+class Car(var position: Int = DEFAULT_POSITION) {
```
외부에서 객체 내부의 값을 변경할 수 있다면 어떤 일들이 생길 수 있을까요?

이를 해결하기 위해서는 어떤 방법들 [1](https://kotlinlang.org/docs/properties.html#getters-and-setters) [2](https://discuss.kotlinlang.org/t/private-setter-for-var-in-primary-constructor/3640/9)이 있을까요?

ddaaac commented on src/main/kotlin/week1/step3/Car.kt:
```diff
@@ -0,0 +1,17 @@
+package week1.step3
+
+class Car(var position: Int = DEFAULT_POSITION) {
+    fun move(randomNumber: Int) {
+        when (randomNumber) {
+            in STOPPING_NUMBERS -> return
+            in FORWARDING_NUMBERS -> position++
+            else -> throw IllegalArgumentException("0-9 사이의 randomNumber 가 아닙니다: randomNumber=$randomNumber")
+        }
```
자동차를 움직이는 방법에는 다양한 방법이 있을 것 같아요.
car.move()
car.position += xx
car.move(xx)

등등 의 여러 방법 중에서 영근님이 이 방법을 사용해주신 이유가 궁금합니다!

y2gcoder replied:
제가 생각했을 때는 책임의 관점에서 생각해보고 car.move()를 사용했던 것 같습니다!

> 전진하는 조건은 0에서 9 사이에서 무작위 값을 구한 후 무작위 값이 4 이상일 경우이다.

위의 기능 요구사항이 차가 전진하는 것에 관련된 요구사항이었습니다. 
여기서 무작위 값을 구하는 것이 Car의 책임으로 보이지는 않았습니다. 

> Car는 임의의 값이 4이상이면 전진하면 된다

까지가 Car가 은닉하고 구현해야 할 사항이라고 생각해서 car.move(xx)를 사용하게 되었습니다. 

car.position += xx 와 같은 형태는 위의 전진하는 조건과 비즈니스 로직이 외부로 노출되어 캡슐화가 지켜지지 않는 코드라 생각해서 배제했습니다. 

그리고 추가로 임의의 값을 만드는 것은 외부로 빼냈을 때 Car의 기능을 단위 테스트로 만들기 더 용이하다고 판단한 것도 추가적인 이유입니다! 이에 대해서는 해당 [링크](https://jojoldu.tistory.com/674) 를 참고했습니다! 



감사합니다!

ddaaac replied:
크으... 너무 이유가 상세하고 명쾌하네요 👍 
저도 완전 동의합니다.

+) 이미 이번 미션에서 다루는 주제는 모두 고려한 것 같아서, 개인적으로 제가 미션을 진행하면서 들었던 생각을 공유해볼게요.
과연 요구사항에 대한 충분한 이해없이 이 시그니처만 봤을 때 클라이언트 입장에서 이해하기 쉬운가?입니다.
`car.move(randomNumber)`라는 것만 놓고보면 randomNumber가 의미하는 바가 명확히 와닿을 수 있지 않겠다는 생각이 들더라구요.
그래서 든 생각은 `car.move(전진을 결정해주는 인터페이스)`, 또는 `Car(engine).move()` 이런식으로 구현을 해보는 것도 좋을 것 같더라구요. 이번 미션에서는 조금 지나친 설계라는 생각이 들 수도 있구요.
영근님의 생각은 어떠신가요?

ddaaac commented on src/main/kotlin/week1/step3/RacingGame.kt:
```diff
@@ -0,0 +1,43 @@
+
+class RacingGame(
+    carCount: Int,
+    private val randomNumber: () -> Int = { Random.nextInt(0, 10) },
```
[functional interface](https://kotlinlang.org/docs/fun-interfaces.html)와 단순 함수형 선언에는 어떤 차이가 있을까요?

y2gcoder replied:
아.. 제가 굳이굳이 람다형 함수를 만들어 사용하고 있었네요!
가르쳐주셔서 감사합니다!

ddaaac commented on src/main/kotlin/week1/step3/RacingGame.kt:
```diff
@@ -0,0 +1,43 @@
+
+    init {
+        validateCarCount(carCount)
+        cars = List(carCount) { Car() }
```
kotlin에서 [리스트 생성](https://www.baeldung.com/kotlin/initialize-list)엔 여러 방법이 있습니다.
적당한 방법을 잘 사용해주신 것 같네요!

ddaaac commented on src/main/kotlin/week1/step3/RacingGame.kt:
```diff
@@ -0,0 +1,43 @@
+        cars = List(carCount) { Car() }
+    }
+
+    private fun validateCarCount(carCount: Int) {
+        if (carCount <= 0) {
+            throw IllegalArgumentException("레이싱 게임을 하기 위해서는 차가 1대 이상이어야 합니다.")
+        }
```
이렇게 carCount를 Int로 받고 RacingGame에서 검증하는 방법도 있지만
[value class](https://kotlinlang.org/docs/inline-classes.html)를 사용하는 방법도 있습니다.
적당한 방법을 잘 골라주시면 될 것 같아요.

ddaaac commented on src/test/kotlin/week1/step3/CarTest.kt:
```diff
@@ -0,0 +1,41 @@
+
+class CarTest : StringSpec({
+    "Car는 4~9의 무작위 값을 받으면 한 칸 전진한다" {
+        listOf(4, 5, 6, 7, 8, 9).forEach { randomNumber ->
+            val sut = Car()
+            val initialPosition = sut.position
+
+            sut.move(randomNumber)
+
+            val currentPosition = sut.position
+            currentPosition shouldBe initialPosition + 1
+        }
+    }
+
+    "Car는 0~3의 무작위 값을 받으면 그대로 멈춘다" {
+        listOf(0, 1, 2, 3).forEach { randomNumber ->
+            val sut = Car()
+            val initialPosition = sut.position
+
+            sut.move(randomNumber)
+
+            val currentPosition = sut.position
+            currentPosition shouldBe initialPosition
+        }
+    }
+
+    "Car는 0~9 이외의 값을 받으면 InvalidArgumentException 예외를 던진다" {
+        listOf(-1, 10).forEach { randomNumber ->
```
꼼꼼한 테스트 케이스와 경계값 테스트까지 깔끔하게 해주셨네요!

ddaaac commented on src/test/kotlin/week1/step3/RacingGameTest.kt:
```diff
@@ -0,0 +1,39 @@
+
+    "RacingGame의 레이스를 실행하면 각 라운드마다 Car들의 위치를 표시해놓은 결과를 받을 수 있다" {
+        val carCount = 3
+        val sut = RacingGame(carCount) { 4 } // 무조건 전진
```
`RacingGame(carCount, alwaysMove)` 처럼 변수를 미리 만들어서 사용한다면 주석 없이도 표현이 가능할 것 같아요! 테스트 간에 재활용도 가능할 것 같구요.

ddaaac reviewed (changes requested):
안녕하세요 영근님!

구현을 깔끔하게 잘 해주셨네요~

몇 가지 코멘트 남겼으니 확인해주시고, 궁금한 부분은 언제든 dm으로 문의주세요!

y2gcoder committed:
docs(step3-racingcar): Add reviewer's feedback

리뷰어님 피드백을 투두로 추가

y2gcoder committed:
feature(step3-racingcar): Make the setter of the Car's position property private

리뷰어님 피드백 1 반영: Car 의 position 프로퍼티를 외부에서 변경하지 못하도록 private set으로 setter 노출 제한

y2gcoder committed:
docs(step3-racingcar): Make reviewer's feedback 2 Checked

리뷰어님 피드백 2 체크: car.move(randomNumber: Int) 를 사용한 이유에 대한 설명 코멘트로 남기기

y2gcoder committed:
docs(step3-racingcar): Create inline value class(CarRoster)

리뷰어님 피드백 5: Inline Value class(CarRoster) 추가

y2gcoder committed:
docs(step3-racingcar): Make reviewer's feedback 4 Checked

리뷰어님 피드백 4: Kotlin List 생성 방법 숙지

y2gcoder committed:
feature(step3-racingcar): Apply Kotlin functional interface to randomNumberGenerator in RacingGame

리뷰어님 피드백 3: RacingGame 생성시 무작위 숫자 생성기에 kotlin functional interface 를 사용
명확한 이름을 사용할 수 있고 자바의 functional interface와 상호 운용이 가능하다는 장점이 있음.

리뷰어님 피드백 6: RacingGameTest 에서 테스트를 위해 만든 RandomNumberGenerator를 기능에 맞는 명확한 이름으로 변경하고 재사용하도록 밖으로 추출

ddaaac reviewed (approved):
안녕하세요 영근님 👋 

리뷰 반영 너무 잘 해주셨네요!

코멘트 하나 남겼으니 확인 부탁드려요.
다음 단계 진행해주세요~
